### PR TITLE
Fix the parsing of span-id, and do it always.

### PR DESCRIPTION
**Pull request by csilvers**
## Summary:
I have some more clarity on what's going on here -- x-cloud-trace-context always has decimal span-ids, so we need to convert to hex always.  And we need to convert all the numbers, not just half of them!  Int/uint ugh.

Issue: https://console.cloud.google.com/support/cases/detail/v2/54166787?project=khan-academy

## Test plan:
Fingers crossed, but I did look at a bunch of log messages to convince myself this is the right thing to do, in addition to what Google said in the support message.

Author: csilvers

Reviewers: MiguelCastillo

Required Reviewers:

Approved By: MiguelCastillo

Checks:

Pull Request URL: https://github.com/Khan/stackdriver-gae-logrus-plugin/pull/2

### CODE REVIEW

**Comment by andyfusniak** [comment]
I want to make sure I've understood this correctly before merging.

1. Previously, we assumed that `X-Cloud-Trace-Context` could contain either a v1 (decimal) or a v2 (hex) formatted string. Now, it seems that v1 will always be a decimal integer—is that correct?

2. For any given trace, some span IDs may use the newer v2 (hex) format, while others may still use the legacy v1 (decimal) format. Therefore, you’ll be converting v1 span IDs to the v2 (hex) format?

3. Assuming my interpretation of point 2 is accurate, you're confident that the high entropy of v1 span IDs ensures a negligible risk of collision when converting them to hex format?

4. When you mentioned "Int/uint ugh," were you referring to the need to handle v1 Span IDs as unsigned 64-bit numbers and this requires us to use `ParseUint` and `FormatUint` from the `strconv` package, rather than `ParseInt`?

According to your code comments, the current logger won't attempt to support v2 at this point, as it ideally should be updated to check for `Traceparent` headers instead of `X-Cloud-Trace-Context` altogether, as per [protocols for context propagation](https://cloud.google.com/trace/docs/trace-context).

I reviewed the [sample code you referenced](https://github.com/googleapis/google-cloud-go/blob/f072178f6fd90537a5782395f4229e4c8b30af7e/logging/logging.go#L834-L855), but noticed it's not exported from the package. I wondered if the OpenTelemetry package might have a utility for parsing these headers. Although I’m not familiar with it, I’m noting here that the [Extract method](https://github.com/open-telemetry/opentelemetry-go/blob/main/propagation/trace_context.go#L72) could be worth exploring. Introducing this dependency might be acceptable in this instance to avoid maintaining custom parsing code.

For a non-Logrus approach, the Go standard library’s `slog` package might be an option, with an [example of its usage here](https://cloud.google.com/trace/docs/setup/go-ot).

**Comment by csilvers** [comment]
Here's how I would summarize it:

1. The requestlogs in appengine (emitted by Google, that users have no control over), used to record the span-id as a base-10 integer.  A few months ago, that code changed to recording the span-id as a base-16 integer.  (Because it updated from v1 of the trace-format to v2.)
2. The X-Cloud-Trace-Context header that google sets continued to set the span-id as a base-10 number.  I thought they were changing it to be base-16, but I was mistaken.  Hence the fix in #2 wasn't really correct.  I do a more correct fix here.
3. The fix that Google wants us to do -- which I did not do here -- is to stop using X-Cloud-Trace-Context entirely, and instead parse the `Traceparent` header.  This header has been v2 since it was created, and thus reports the span-id as a base-16 number.

Does that answer your questions?

> I reviewed the [sample code you referenced](https://github.com/googleapis/google-cloud-go/blob/f072178f6fd90537a5782395f4229e4c8b30af7e/logging/logging.go#L834-L855), but noticed it's not exported from the package

Yeah, I showed it just to demonstrate we're doing the same thing they are, not that we can use their code directly (though it would be nice if we could!)

**Comment by csilvers** [comment]
> When you mentioned "Int/uint ugh," were you referring to the need to handle v1 Span IDs as unsigned 64-bit numbers and this requires us to use ParseUint and FormatUint from the strconv package, rather than ParseInt?

That's correct!

**Comment by andyfusniak** [comment]
Thanks for the background info. Yes, that covers my questions and helps my understanding. It took me longer than usual to get my head around because I've not used App Engine or Cloud Run for a long time. I've been going solely off of the docs and our discussions.

I just setup a dummy project to test it out. You're probably familiar with the following, so the explanation is mostly for my future understanding.

I viewed a request by its trace ID:
Screenshot from 2024-11-14 20-35-46

When I expanded the JSON data in the console there are no JSON properties shown for the `traceparent`.
Screenshot from 2024-11-14 20-36-30

However, I output the request headers in my Go HTTP service and the `traceparent` header is being sent by App Engine (standard environment).

Screenshot from 2024-11-14 20-47-52


Note also that the span ID (decimal in red) from the `X-Cloud-Trace-Context` header and span ID (hex in orange) from the `Traceparent` header both match (checked those in Python).

```
>>> hex(262214736119705593)
'0x3a392e9c949aff9'
```
Screenshot from 2024-11-14 20-30-12

Screenshot from 2024-11-14 20-31-59

I did notice the `spanId` sometimes came out as decimals from the Logrus formatter (without the patch). My only concern was that if everything was coming out hex formatted that it might break someone relying on the decimal format, although this seems highly unlikely.

In summary

- I will merge the PR.
- I'll look at a solution in near future to use the `Traceparent` header and fall back to the `X-Cloud-Trace-Context` it not there.
- I'll probably just parse the headers myself to avoid the dependency to the footprint HTTP services. (I originally wrote this for a monolith I was working on.  I expect you're using Microservices since you're focused on Trace IDs, span IDs and potentially parent span IDs forming a tree, so keeping the dependencies might be better). 

Let me know if I've missed anything.

**Comment by csilvers** [comment]
That seems great!
